Sources/Commands/SwiftPackageTool.swift

```diff
@@ -687,6 +728,236 @@ extension SwiftPackageTool {
+                                    checksum: checksum)
+            default:
+                swiftTool.diagnostics.emit(.error("unsupported target type '\(type)'; supported types are library, executable, test, and binary"))
+                throw ExitCode.failure
+            }
+
+            do {
+                try editor.addTarget(newTarget,
+                                     productPackageNameMapping: createProductPackageNameMapping(swiftTool: swiftTool))
+            } catch Diagnostics.fatalError {
+                throw ExitCode.failure
+            }
+        }
+
+        private func createProductPackageNameMapping(swiftTool: SwiftTool) throws -> [String: String] {
+            let graph = try swiftTool.loadPackageGraph()
+            var productPackageNameMapping: [String: String] = [:]
+            for dependencyPackage in graph.rootPackages.flatMap(\.dependencies) {
+                for product in dependencyPackage.products {
+                    productPackageNameMapping[product.name] = dependencyPackage.name
+                }
+            }
+            return productPackageNameMapping
+        }
+
+        private func verifyNoTargetBinaryOptionsPassed(swiftTool: SwiftTool) throws {
+            guard url == nil else {
+                swiftTool.diagnostics.emit(.error("option '--url' is only supported for binary targets"))
+                throw ExitCode.failure
+            }
+            guard path == nil else {
+                swiftTool.diagnostics.emit(.error("option '--path' is only supported for binary targets"))
+                throw ExitCode.failure
+            }
+            guard checksum == nil else {
+                swiftTool.diagnostics.emit(.error("option '--checksum' is only supported for binary targets"))
+                throw ExitCode.failure
+            }
+        }
+    }
+
+    struct AddProduct: SwiftCommand {
+        static let configuration = CommandConfiguration(
+            abstract: "Add a product to the current package.")
+
+        @OptionGroup()
+        var swiftOptions: SwiftToolOptions
+
+        @Argument(help: "The name of the new product")
+        var name: String
+
+        @Option(help: "The type of the new product (library, static-library, dynamic-library, or executable)")
+        var type: ProductType?
+
+        @Option(help: "A comma-separated list of target names")
+        var targets: String
+
+        func run(_ swiftTool: SwiftTool) throws {
+            let packageRoot = try swiftTool.getPackageRoot()
+            let editor = try PackageEditor(manifestPath: packageRoot.appending(component: Manifest.filename),
+                                           repositoryManager: swiftTool.getActiveWorkspace().repositoryManager,
+                                           toolchain: swiftTool.getToolchain(),
+                                           diagnosticsEngine: swiftTool.diagnostics)
+            let targets = self.targets.split(separator: ",").map(String.init)
```

Comment:
This is related to the [comment](https://forums.swift.org/t/se-0301-package-editor-commands/44447/13) I left in the review thread, but a bit more detailed code-wise.

For`AddTarget.dependencies` and `AddProduct.targets`, instead of separating a comma-separated string, I think it might be better to use arrays directly with Swift Argument Parser's [`ArrayParsingStrategy.upToNextOption`](https://github.com/apple/swift-argument-parser/blob/75c4dcd2e7cd0878aa1d29e6b493a8abcd65d753/Sources/ArgumentParser/Parsable%20Properties/Option.swift#L247):

```swift
struct AddTarget {
    // ...
    @Option(
        parsing: .upToNextOption,
        help: "A space-separated list of target dependency names (targets and/or dependency products)"
    )
    var dependencies: [String]?
    // ...
}

struct AddProduct {
    // ...
    @Option(
        parsing: .upToNextOption,
        help: "A space-separated list of target names"
    )
    var targets: [String]
    // ...
}
```

Reply:
Thanks, I responded to your comment on the forums. Spaces is probably a little better but I want to think about it for a day or two before I make any revisions